### PR TITLE
feat(control): add new dump route config api for control api

### PR DESCRIPTION
**Pull request by Zheaoli**
### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [ ] Have you modified the corresponding document?
* [ ] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

Add new API for dumping config in control api

### CODE REVIEW

File: apisix/control/v1.lua [hunk 1]
```diff
@@ -189,4 +212,10 @@ return {
         uris = {"/gc"},
         handler = _M.trigger_gc,
     },
+    -- /v1/dump
```
Comment (spacewander) [inline apisix/control/v1.lua:215]: Please use restful like API names for dump specific resources, like `/v1/upstreams/xxx` and `/v1/upstreams`.

File: apisix/control/v1.lua [hunk 2]
```diff
@@ -156,6 +157,28 @@ function _M.get_health_checker()
     return 200, info
 end
 
+local function iter_add_get_routes_info(infos, values)
+    local temp_infos = {}
+    for _, route in core.config_util.iterate_values(values) do
+        local new_route = copy.deepcopy(route)
```
Comment (spacewander) [inline apisix/control/v1.lua:163]: We already have `core.table.deepcopy`

File: docs/en/latest/control-api.md [hunk 3]
```diff
@@ -204,3 +204,160 @@ Introduced since `v2.8`.
...
+[
+  {
+    "value": {
+      "uris": [
+        "/hello"
+      ],
+      "id": "1",
+      "upstream": {
+        "scheme": "http",
+        "checks": {
```
Comment (spacewander) [inline docs/en/latest/control-api.md:224]: The example here is too long to focus on the point. Can we remove the healthcheck part?

File: t/control/routes.t [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+    location /t {
+        content_by_lua_block {
+            local json = require("toolkit.json")
+            local t = require("lib.test_admin")
+            local code, body, res = t.test('/v1/routes',
+                ngx.HTTP_GET)
+            res = json.decode(res)
+            if res[1] then
+                local data = {}
+                data.uris = res[1].value.uris
```
Comment (spacewander) [inline t/control/routes.t:81]: We should also check the upstream part

File: t/control/routes.t [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+  -
+    id: 1
+    uris:
+        - /hello
+    upstream:
+      nodes:
+        "127.0.0.1:1980": 1
+        "127.0.0.1:1988": 1
+      type: roundrobin
+      checks:
```
Comment (spacewander) [inline t/control/routes.t:146]: We don't need to define a long upstream for a test just check if the route exists

File: apisix/control/v1.lua [hunk 6]
```diff
@@ -156,6 +156,40 @@ function _M.get_health_checker()
     return 200, info
 end
 
+local function iter_add_get_routes_info(infos, values, route_id)
```
Comment (spacewander) [inline apisix/control/v1.lua:159]: The infos can be created inside this function, and there is no need to pass out a flag -- use nil if route_id given but not matched.

File: docs/en/latest/control-api.md (hunk 3 above)
Comment (Zheaoli) [reply docs/en/latest/control-api.md:224]: of course

File: apisix/control/v1.lua [hunk 7]
```diff
@@ -156,6 +156,42 @@ function _M.get_health_checker()
...
+        end
+        core.table.insert(infos, new_route)
+        if route_id and route.value.id == route_id then
+            return new_route, true
+        end
+    end
+    if not route_id then
+        return infos, true
+    end
+    return nil, false
```
Comment (spacewander) [inline apisix/control/v1.lua:174]: We don't need a flag, just check the first result is nil or not

File: t/control/routes.t (hunk 4 above)
Comment (spacewander) [reply t/control/routes.t:70]: Need to check the upstream part too

**Review by spacewander — APPROVED** [review]

**Review by tzssangglass — APPROVED** [review]

**Pull request merged**